fendermint/vm/topdown/src/vote/store.rs

```diff
@@ -0,0 +1,187 @@
+    fn earliest_vote_height(&self) -> Result<Option<BlockHeight>, Error>;
+
+    /// Get the latest block height of the votes stored
+    fn latest_vote_height(&self) -> Result<Option<BlockHeight>, Error>;
+
+    /// Store the vote at the target height
+    fn store_vote(&mut self, height: BlockHeight, vote: Vote) -> Result<(), Error>;
+
+    /// Checks if the validator has voted at the target block height
+    fn has_voted(&self, height: &BlockHeight, validator: &ValidatorKey) -> Result<bool, Error>;
+
+    /// Get the votes at the height.
+    fn get_votes_at_height(&self, height: BlockHeight) -> Result<VoteAgg, Error>;
+
+    /// Purge all votes at the specific height. This could be the target height has reached a
+    /// quorum and the history is not needed.
+    fn purge_votes_at_height(&mut self, height: BlockHeight) -> Result<(), Error>;
+}
+
+#[derive(Default)]
+pub struct InMemoryVoteStore {
+    votes: BTreeMap<BlockHeight, HashMap<ValidatorKey, Vote>>,
+}
+
+impl VoteStore for InMemoryVoteStore {
+    fn earliest_vote_height(&self) -> Result<Option<BlockHeight>, Error> {
+        Ok(self.votes.first_key_value().map(|(k, _)| *k))
+    }
+
+    fn latest_vote_height(&self) -> Result<Option<BlockHeight>, Error> {
+        Ok(self.votes.last_key_value().map(|(k, _)| *k))
+    }
+
+    fn store_vote(&mut self, height: BlockHeight, vote: Vote) -> Result<(), Error> {
+        match self.votes.entry(height) {
+            Entry::Vacant(_) => {
+                let mut map = HashMap::new();
+                map.insert(vote.voter(), vote);
+                self.votes.insert(height, map);
+            }
+            Entry::Occupied(mut v) => {
+                let key = vote.voter();
+                v.get_mut().insert(key, vote);
+            }
+        }
+        Ok(())
+    }
+
+    fn has_voted(&self, height: &BlockHeight, validator: &ValidatorKey) -> Result<bool, Error> {
+        let Some(votes) = self.votes.get(height) else {
+            return Ok(false);
+        };
+        Ok(votes.contains_key(validator))
+    }
+
+    fn get_votes_at_height(&self, height: BlockHeight) -> Result<VoteAgg, Error> {
+        let votes = self
+            .votes
+            .get(&height)
+            .map(|v| v.values().collect())
+            .unwrap_or_default();
+        Ok(VoteAgg::new(votes))
+    }
+
+    fn purge_votes_at_height(&mut self, height: BlockHeight) -> Result<(), Error> {
+        self.votes.remove(&height);
+        Ok(())
+    }
+}
+
+/// The aggregated votes  from different validators.
+pub struct VoteAgg<'a>(Vec<&'a Vote>);
+
+impl<'a> VoteAgg<'a> {
+    pub fn new(votes: Vec<&'a Vote>) -> Self {
+        Self(votes)
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    pub fn into_owned(self) -> Vec<Vote> {
+        self.0.into_iter().cloned().collect()
+    }
+
+    pub fn observation_weights(&self, power_table: &PowerTable) -> Vec<(&Observation, Weight)> {
```

Comment:
Better to make the returned list ordered by the weight, so you won't need to iterate over it for to find a `> quorum_threshold` item.

Reply:
I just returned the max weight that's what we care really.